Python/ast_opt.c

```diff
@@ -452,12 +452,51 @@ static int astfold_excepthandler(excepthandler_ty node_, PyArena *ctx_, int opti
     } \
 }
 
+static int
+isdocstring(stmt_ty s)
+{
+    if (s->kind != Expr_kind)
+        return 0;
+    if (s->v.Expr.value->kind == Str_kind)
+        return 1;
+    if (s->v.Expr.value->kind == Constant_kind)
+        return PyUnicode_CheckExact(s->v.Expr.value->v.Constant.value);
+    return 0;
+}
+
+static int
+astfold_body(asdl_seq *stmts, PyArena *ctx_, int optimize_)
+{
+    if (!asdl_seq_LEN(stmts)) {
+        return 1;
+    }
+    int docstring = isdocstring((stmt_ty)asdl_seq_GET(stmts, 0));
+    CALL_SEQ(astfold_stmt, stmt_ty, stmts);
```

Comment:
Loop from 1 seems simpler than JoinedStr hack.

```c
    for (int i = docstring; i < asdl_seq_LEN(stmts); i++) {
        stmt_ty st = (stmt_ty)asdl_seq_GET(stmts, i);
        if (st != NULL && !astfold_stmt(st, ctx_, optimize_)) {
            return 0;
        }
    }
```

Reply:
A JoinedStr hack is needed for preventing interpreting a new constant string created by optimization as a docstring:
```python
def foo():
    'Not a' + ' docstring'
    pass
```
but we still want to apply optimizations to the first statement.